### PR TITLE
feat: Migrate storage from S3 native to S3-compatible B2 with CloudFlare proxy

### LINKED ISSUE #163: Migrate all assets off of AWS S3

### PR DESCRIPTION
**Pull request by kanyesthaker**
### Description of Changes
- Migrate from AWS S3 to a BackBlaze B2 always-hot cloud storage bucket
- Set up a read-only proxy via CloudFlare for free egress
- Documentation and installation guide available at this [Notion page](https://www.notion.so/irisapp/Using-B2-Cloud-Storage-fac73aa9e3d64ffd9b30e2e2b2d2aa51) 
- Reformat "resolveStoragePath" function

In order to set up the CloudFlare proxy webworker, there is a new module `storage-api` with its own dependencies. The `worker.ts` file defines the actual webworker functionality by intercepting the `fetch` function, using the B2 S3-compatible API to pre-sign the request (read more [here](https://docs.aws.amazon.com/AmazonS3/latest/userguide/ShareObjectPreSignedURL.html)), and then fetching the object. The `wrangler.toml` file defines metadata and environment variables referenced by the webworker. Private variables (such as the true API key) are directly maintained and encrypted in the CloudFlare console

### Link to Issue
closes #163 

### Testing Procedure
World loads properly!
image

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -20,6 +20,7 @@
     "@types/react-fontawesome": "^1.6.5",
     "@types/uuid": "^9.0.0",
     "@types/ws": "^8.5.3",
+    "aws4fetch": "^1.0.17",
```
Comment (samjgorman) [inline package.json:23]: What is this for? 

Read about the repo and its described as an AWS client. If we're no longer using AWS what role does this play?
https://github.com/mhart/aws4fetch

File: src/common/constants.ts [hunk 2]
```diff
@@ -77,21 +77,21 @@ export abstract class Boundary {
...
-    resolveS3Path('calibration_skybox/positive-y.png', 'texture'),
-    resolveS3Path('calibration_skybox/negative-y.png', 'texture'),
-    resolveS3Path('calibration_skybox/positive-x.png', 'texture'),
-    resolveS3Path('calibration_skybox/negative-x.png', 'texture'),
+    resolveStoragePath('calibration_skybox/negative-z.png', Bucket.Texture),
+    resolveStoragePath('calibration_skybox/positive-z.png', Bucket.Texture),
+    resolveStoragePath('calibration_skybox/positive-y.png', Bucket.Texture),
+    resolveStoragePath('calibration_skybox/negative-y.png', Bucket.Texture),
+    resolveStoragePath('calibration_skybox/positive-x.png', Bucket.Texture),
+    resolveStoragePath('calibration_skybox/negative-x.png', Bucket.Texture),
```
Comment (samjgorman) [inline src/common/constants.ts:94]: Nice cleanup

File: src/common/utils/io.ts [hunk 3]
```diff
@@ -1,21 +1,18 @@
+export enum Bucket {
+  Texture = 'textures/',
+  Model = 'models/',
+  Terrain = 'terrain/',
+}
+
 /**
  * Converts a shorthand file path to a full path for an S3 bucket.
```
Comment (samjgorman) [inline src/common/utils/io.ts:8]: Update docstring

File: storage-api/worker.ts [hunk 4]
```diff
@@ -0,0 +1,110 @@
+// @ts-nocheck
+// Proxy Backblaze S3 compatible API requests, sending notifications to a webhook
```
Comment (samjgorman) [inline storage-api/worker.ts:2]: I can use some help understanding why we need this.  Why are we proxying S3 compatible requests? For our migration?

File: storage-api/worker.ts [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+  // Remove trailing slashes
+  path = path.replace(/\/$/, '');
+  // Split the path into segments
+  const pathSegments = path.split('/');
+
+  if (ALLOW_LIST_BUCKET !== 'true') {
+    // Don't allow list bucket requests
+    if (
+      (BUCKET_NAME === '$path' && pathSegments[0].length < 2) || // https://endpoint/bucket-name/
+      (BUCKET_NAME !== '$path' && path.length === 0)
```
Comment (samjgorman) [inline storage-api/worker.ts:44]: Readability is taking a hit with these conditionals... Would recommend cleaning this up

**Review by samjgorman — COMMENTED** [review]
Nice work moving quickly on this & taking initiative. Holding off on approving until I have a better understanding of what we're doing here, especially understanding the relationship of the Cloudflare web worker to Backblaze in order to avoid egress fees.  @kanyesthaker and I have already started to sync off Github & left a few relevant questions here

File: package.json (hunk 1 above)
Comment (kanyesthaker) [reply package.json:23]: BackBlaze uses an S3-compatible API, which means that we can use the AWS S3 API in order to interact with BackBlaze directly. This AWS Client allows us to authenticate requests made by the CloudFlare worker.

File: src/common/utils/io.ts (hunk 3 above)
Comment (kanyesthaker) [reply src/common/utils/io.ts:8]: good catch

File: storage-api/worker.ts (hunk 4 above)
Comment (kanyesthaker) [reply storage-api/worker.ts:2]: Answered async^

File: storage-api/worker.ts (hunk 5 above)
Comment (kanyesthaker) [reply storage-api/worker.ts:44]: good point -- this is code ripped from https://github.com/backblaze-b2-samples/cloudflare-b2 -- doesn't hurt to clean it up though

File: src/common/constants.ts [hunk 6]
```diff
@@ -77,21 +77,21 @@ export abstract class Boundary {
...
-    resolveS3Path('posy.jpg', 'texture'),
-    resolveS3Path('negy.jpg', 'texture'),
-    resolveS3Path('posx.jpg', 'texture'),
-    resolveS3Path('negx.jpg', 'texture'),
+    resolveStoragePath('negz.jpg', Bucket.Texture),
+    resolveStoragePath('posz.jpg', Bucket.Texture),
+    resolveStoragePath('posy.jpg', Bucket.Texture),
+    resolveStoragePath('negy.jpg', Bucket.Texture),
+    resolveStoragePath('posx.jpg', Bucket.Texture),
+    resolveStoragePath('negx.jpg', Bucket.Texture),
```
Comment (shalins) [inline src/common/constants.ts:85]: Can `resolveStoragePath` not just take in a list to avoid copy-paste?

**Review by shalins — APPROVED** [review]
Approving, but just the small nitpick about overloading `resolveStoragePath` take in either a string or a list of strings to avoid code dup. Also, would you mind transpiling your Notion page into a README and placing in the storage-api dir? That would make it super easy to reference in the future I think. Apart from that, great work as always!

File: src/common/utils/io.ts (hunk 3 above)
Comment (samjgorman) [reply src/common/utils/io.ts:8]: Doesn't look like this was resolved

File: storage-api/README.md [hunk 7]
```diff
@@ -0,0 +1,3 @@
+# SpaceBar Storage API
```
Comment (samjgorman) [inline storage-api/README.md:1]: Clear & concise!

**Review by samjgorman — APPROVED** [review]
LGTM! Thanks for this! Left one non-blocking nit

**Pull request merged**